README.md
# tiled_tab_indicator / Tiled Tab Indicator

TiledTabIndicator is a tab indicator with multiple colored tiles in a line.

<img src="https://media.giphy.com/media/QUIefume8b1nSpKHBa/giphy.gif"/>

## Getting Started

Add this to your package's 'pubspec.yaml' file:

```
´´´
dependencies:
    tiled_tab_indicator: ^1.0.1
```

Now in your Dart code, you can import like this:

```
import 'package:tiled_tab_indicator/tiled_tab_indicator.dart';
```

# Usage

You can use it just like the UnderlineTabIndicator:

```dart
    TabBar(
        indicator: TiledTabIndicator(
            indicatorHeight: 5,
            colors: [
                Color(0xff655dfc),
                Color(0xff5174fa),
                Color(0xff3299fc),
                Color(0xff3dc9de),
            ],
            style: TiledIndicatorStyle.normal, // or .round
        ),
        indicatorSize: TabBarIndicatorSize.tab, // TiledTabIndicator will change size
                                                // just like UnderlineTabIndicator
                                                // if you'd use .label here
    ),
    tabs: [
        Tab(text: "Tab1"),
        Tab(text: "Tab2"),
    ],
    ...
```


CHANGELOG.md
## [1.0.1]

-   Fixed a typo in the example

## [1.0.0]

-   Initial release


test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter_test/flutter_test.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // it sure is empty around here
  });
}


lib/tiled_tab_indicator.dart
library tiled_tab_indicator;

import 'package:flutter/widgets.dart';

/// The styling of the indicator
enum TiledIndicatorStyle {
  /// Square Rects
  normal,

  /// Outer edges of the indicator are round
  round,
}

/// A TabIndicator that can have multiple colored tiles.
/// Used with [TabBar.indicator] to draw a rectangles in a line below the
/// selected tab.
@immutable
class TiledTabIndicator extends Decoration {
  /// Height of the [TiledTabIndicator]
  final double indicatorHeight;

  /// Colors to use for the tiles,
  /// colors appear in order, index 0 = first color...
  final List<Color> colors;

  /// Style of the indicator
  /// see [TiledIndicatorStyle] for the options
  final TiledIndicatorStyle style;

  /// Creates an underline style tab indicator with colored tiles.
  ///
  /// There are no colors by default, so if you pass an empty
  /// list, the indicator will not be shown
  ///
  /// By default, the style is [TiledIndicatorStyle.normal]
  const TiledTabIndicator(
      {@required this.indicatorHeight, @required this.colors, this.style})
      : assert(indicatorHeight != null),
        assert(colors != null);

  @override
  _TiledIndicatorPainter createBoxPainter([VoidCallback onChanged]) {
    return new _TiledIndicatorPainter(this, onChanged);
  }
}

class _TiledIndicatorPainter extends BoxPainter {
  final TiledTabIndicator decoration;

  _TiledIndicatorPainter(this.decoration, VoidCallback onChanged)
      : assert(decoration != null),
        super(onChanged);

  @override
  void paint(Canvas canvas, Offset offset, ImageConfiguration configuration) {
    assert(configuration != null);
    assert(configuration.size != null);

    /// internal offset, so the rects don't stack on top of each other
    final off = configuration.size.width / decoration.colors.length;
    // possible rounding for the corners
    final rounding = Radius.circular(decoration.indicatorHeight / 2.0);
    for (int i = 0; i < decoration.colors.length; i++) {
      final Color color = decoration.colors[i];
      final Paint paint = Paint()
        ..color = color
        ..style = PaintingStyle.fill;
      Rect rect = Offset(offset.dx + off * i,
              configuration.size.height - decoration.indicatorHeight) &
          Size(configuration.size.width / decoration.colors.length,
              decoration.indicatorHeight);
      // styling
      switch (decoration.style ?? TiledIndicatorStyle.normal) {
        case TiledIndicatorStyle.normal:
          canvas.drawRect(rect, paint);
          break;
        case TiledIndicatorStyle.round:
          // only round the outer corners
          if (decoration.colors.length == 1) {
            // length of 1 needs both left and right corners rounded
            canvas.drawRRect(
                RRect.fromRectAndCorners(
                  rect,
                  topLeft: rounding,
                  bottomLeft: rounding,
                  topRight: rounding,
                  bottomRight: rounding,
                ),
                paint);
          } else if (i == 0) {
            // if length > 1, index 0 should only round left corners
            canvas.drawRRect(
                RRect.fromRectAndCorners(
                  rect,
                  topLeft: rounding,
                  bottomLeft: rounding,
                ),
                paint);
          } else if (i == decoration.colors.length - 1) {
            // if length > 1, last index should only round right corners
            canvas.drawRRect(
                RRect.fromRectAndCorners(
                  rect,
                  topRight: rounding,
                  bottomRight: rounding,
                ),
                paint);
          } else {
            // if length > 1, indexes in the middle should not round corners
            canvas.drawRect(rect, paint);
          }

          break;
      }
    }
  }
}


example/main.dart
import 'package:flutter/material.dart';
import 'package:tiled_tab_indicator/tiled_tab_indicator.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
        // This makes the visual density adapt to the platform that you run
        // the app on. For desktop platforms, the controls will be smaller and
        // closer together (more dense) than on mobile platforms.
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return DefaultTabController(
      length: 3,
      child: Scaffold(
        appBar: AppBar(
          centerTitle: true,
          title: Text(
            "Tiled Tab Indicator",
            style: TextStyle(color: Colors.black),
          ),
          backgroundColor: Colors.white,
          bottom: TabBar(
            labelStyle: TextStyle(fontWeight: FontWeight.w700),
            labelColor: Colors.black,
            unselectedLabelColor: Colors.black26,
            indicator: TiledTabIndicator(
              indicatorHeight: 5,
              colors: [
                Color(0xff655dfc),
                Color(0xff5174fa),
                Color(0xff3299fc),
                Color(0xff3dc9de),
              ],
              style: TiledIndicatorStyle.normal,
            ),
            tabs: [
              Tab(text: "Home"),
              Tab(text: "Info"),
              Tab(text: "Data"),
            ],
          ),
        ),
        body: TabBarView(children: [
          Center(child: Text("Home tab")),
          Center(child: Text("Info tab")),
          Center(child: Text("Data tab")),
        ]),
      ),
    );
  }
}
